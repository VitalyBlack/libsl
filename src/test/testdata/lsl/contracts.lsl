libsl "1.0.0";
library simple;

types {
   Int(int);
}

automaton A {
   var i: Int;

   fun f(param: Int)
   requires test1: param >= i;
   ensures: param' < param;
   {

   }
}