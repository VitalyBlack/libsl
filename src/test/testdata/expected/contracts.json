{
  "name": "simple",
  "lslVersion": "1.0.0",
  "types": [
    {
      "name": "Int",
      "isPointer": false,
      "kind": "simple",
      "realType": {
        "name": "int",
        "isPointer": false,
        "kind": "real"
      }
    }
  ],
  "automata": [
    {
      "name": "A",
      "type": "Int",
      "constructorVariables": [],
      "variables": [
        {
          "name": "i",
          "type": "Int"
        }
      ],
      "states": [],
      "shifts": [],
      "functions": [
        {
          "name": "f",
          "automaton": "A",
          "args": [
            {
              "name": "param",
              "type": "Int"
            }
          ],
          "contracts": [
            {
              "name": "test1",
              "kind": "REQUIRES",
              "expression": {
                "kind": "binary",
                "op": "GT_EQ",
                "left": {
                  "kind": "qualifiedAccess",
                  "access": {
                    "kind": "variableAccess",
                    "name": "param",
                    "variableInfo": {
                      "name": "param",
                      "fullName": "f.param",
                      "type": {
                        "name": "Int",
                        "isPointer": false,
                        "kind": "simple",
                        "realType": {
                          "name": "int",
                          "isPointer": false,
                          "kind": "real"
                        }
                      },
                      "kind": "functionArgument",
                      "function": "A.f",
                      "functionArgTypes": [
                        "Int"
                      ]
                    },
                    "type": "Int"
                  }
                },
                "right": {
                  "kind": "qualifiedAccess",
                  "access": {
                    "kind": "variableAccess",
                    "name": "i",
                    "variableInfo": {
                      "name": "i",
                      "fullName": "A.i",
                      "type": {
                        "name": "Int",
                        "isPointer": false,
                        "kind": "simple",
                        "realType": {
                          "name": "int",
                          "isPointer": false,
                          "kind": "real"
                        }
                      },
                      "kind": "automatonVariable",
                      "automaton": "A"
                    },
                    "type": "Int"
                  }
                }
              }
            },
            {
              "kind": "ENSURES",
              "expression": {
                "kind": "binary",
                "op": "LT",
                "left": {
                  "kind": "oldValue",
                  "value": {
                    "kind": "qualifiedAccess",
                    "access": {
                      "kind": "variableAccess",
                      "name": "param",
                      "variableInfo": {
                        "name": "param",
                        "fullName": "f.param",
                        "type": {
                          "name": "Int",
                          "isPointer": false,
                          "kind": "simple",
                          "realType": {
                            "name": "int",
                            "isPointer": false,
                            "kind": "real"
                          }
                        },
                        "kind": "functionArgument",
                        "function": "A.f",
                        "functionArgTypes": [
                          "Int"
                        ]
                      },
                      "type": "Int"
                    }
                  }
                },
                "right": {
                  "kind": "qualifiedAccess",
                  "access": {
                    "kind": "variableAccess",
                    "name": "param",
                    "variableInfo": {
                      "name": "param",
                      "fullName": "f.param",
                      "type": {
                        "name": "Int",
                        "isPointer": false,
                        "kind": "simple",
                        "realType": {
                          "name": "int",
                          "isPointer": false,
                          "kind": "real"
                        }
                      },
                      "kind": "functionArgument",
                      "function": "A.f",
                      "functionArgTypes": [
                        "Int"
                      ]
                    },
                    "type": "Int"
                  }
                }
              }
            }
          ],
          "statements": []
        }
      ]
    }
  ]
}